File: src/main/java/mods/eln/sim/electrical/ElectricalLoad.java
package mods.eln.sim.electrical;

import mods.eln.sim.electrical.mna.component.Bipole;
import mods.eln.sim.electrical.mna.component.Component;
import mods.eln.sim.electrical.mna.component.Line;
import mods.eln.sim.electrical.mna.state.State;
import mods.eln.sim.electrical.mna.state.VoltageStateLineReady;

public class ElectricalLoad extends VoltageStateLineReady {

    public static final State groundLoad = null;

    private double Rs = ElectricalConstants.HIGH_IMPEDANCE;

    public ElectricalLoad() {
    }

    //public VoltageState state = new VoltageState();

    public void setRs(double Rs) {
        if (this.Rs != Rs) {
            this.Rs = Rs;
            for (Component c : getConnectedComponents()) {
                if (c instanceof ElectricalConnection) {
                    ((ElectricalConnection) c).notifyRsChange();
                }
            }
        }
    }

    public double getRs() {
        return Rs;
    }

    public void highImpedance() {
        setRs(ElectricalConstants.HIGH_IMPEDANCE);
    }
//	ArrayList<ElectricalConnection> electricalConnections = new ArrayList<ElectricalConnection>(4);

    public double getI() {
        double i = 0;
        for (Component c : getConnectedComponents()) {
            if (c instanceof Bipole && (!(c instanceof Line)))
                i += Math.abs(((Bipole) c).getCurrent());
        }
        return i * 0.5;
    }

    public double getCurrent() {
        return getI();
    }
}

File: src/main/java/mods/eln/sim/mna/state/ElectricalLoad.kt
package mods.eln.sim.mna.state

import mods.eln.sim.mna.misc.ElectricalConnection
import mods.eln.sim.mna.passive.Bipole
import mods.eln.sim.mna.passive.Line
import mods.eln.sim.mna.misc.MnaConst

open class ElectricalLoad : VoltageStateLineReady() {

    var rs = MnaConst.highImpedance
        set(Rs) {
            if (this.rs != Rs) {
                field = Rs
                for (c in connectedComponents) {
                    if (c is ElectricalConnection) {
                        c.notifyRsChange()
                    }
                }
            }
        }

    val i: Double
        get() {
            var i = 0.0
            for (c in connectedComponents) {
                if (c is Bipole && c !is Line)
                    i += Math.abs(c.getCurrent())
            }
            return i * 0.5
        }

    val current: Double
        get() = i

    fun highImpedance() {
        rs = MnaConst.highImpedance
    }
}
